plugins/renepay/payment.c

```diff
@@ -180,74 +129,71 @@ void payment_assert_delivering_all(const struct payment *p)
 	}
 }
 
-struct command_result *renepay_success(struct renepay * renepay)
+struct command_result *payment_success(struct payment *p)
 {
 	debug_info("calling %s",__PRETTY_FUNCTION__);
-	struct payment *p = renepay->payment;
 
-	payment_success(p);
+	p->status = PAYMENT_SUCCESS;
 	payment_assert_delivering_all(p);
 
+	/* We only finish command once: its destructor clears this. */
+	if (!p->cmd)
+		return NULL;
+
 	struct json_stream *response
-		= jsonrpc_stream_success(renepay->cmd);
+		= jsonrpc_stream_success(p->cmd);
 
 	/* Any one succeeding is success. */
 	json_add_preimage(response, "payment_preimage", p->preimage);
 	json_add_sha256(response, "payment_hash", &p->payment_hash);
 	json_add_timeabs(response, "created_at", p->start_time);
-	json_add_u32(response, "parts", renepay_parts(renepay));
+	json_add_u32(response, "parts", payment_parts(p));
 	json_add_amount_msat(response, "amount_msat",
 				  p->amount);
 	json_add_amount_msat(response, "amount_sent_msat",
 				  p->total_sent);
 	json_add_string(response, "status", "complete");
 	json_add_node_id(response, "destination", &p->destination);
 
-	return command_finished(renepay->cmd, response);
+	return command_finished(p->cmd, response);
 }
 
-struct command_result *renepay_fail(
-	struct renepay * renepay,
+struct command_result *payment_fail(
```

Comment:
I know this is question is addressed in follow-up commit, but just for the record. I have concerns of what happens when a flow fails after the command had finished.

Reply:
Yep: payment is a child of plugin (once we get to the point of actually making payments), so it outlives the cmd.

Any time we free the cmd, the destructor sets payment->cmd to NULL and we check this before trying to access it again.